app/javascript/lib/imperial.js

```diff
@@ -1336,31 +1356,46 @@ export default class Imperial {
         return;
       }
       case 'factory': {
-        // If nation cannot afford to build a factory
-        if (this.nations.get(this.currentNation).treasury < 5) {
-          if (this.variant === 'withoutInvestorCard') {
-            this.roundOfInvestment();
-          } else {
-            this.handleAdvancePlayer();
-            return;
-          }
+        const homeProvinces = new Set();
+        for (const homeProvince of this.board.byNation.get(this.currentNation)) {
+          homeProvinces.add(translateProvinceModel(homeProvince, this.provinces, this.units, this.board));
         }
 
-        for (const province of this.board.byNation.get(action.payload.nation)) {
-          if (
-            !this.provinces.get(province).factory
-            && this.nobodyIsOccupying(province, this.currentNation)
-          ) {
-            this.availableActions.add(Action.buildFactory({ province }));
+        const buildPermissionsUseCase = new FactorySlotBuildPermissions(this.rondel.factorySlot);
+        if (buildPermissionsUseCase.canAffordToBuild(currentNation, currentPlayer)) {
+          this.buildingFactory = true;
+
+          const buildCostsUseCase = new FactorySlotBuildChargeCosts(this.rondel.factorySlot);
+          const nationCosts = buildCostsUseCase.nationCosts(currentNation);
+          const playerCosts = buildCostsUseCase.playerCosts(currentNation, currentPlayer);
+
+          for (const buildableProvince of buildPermissionsUseCase.buildableFactoriesLocations(homeProvinces)) {
+            this.availableActions.add(
+              Action.buildFactory({
+                province: buildableProvince.representation,
+                player: this.currentPlayerName,
+                nationCosts,
+                playerCosts,
+              }),
+            );
           }
+
+          this.availableActions.add(
+            Action.skipBuildFactory({
+              nation: this.currentNation,
+              player: this.currentPlayerName,
+            }),
+          );
+        } else {
+          this.annotatedLog.push(
+            Action.couldNotBuildFactory({
+              nation: this.currentNation,
+            }),
+          );
+
+          this.handlePassingThroughInvestor();
```

Comment:
Is this redundant of line 194? (Although, 194, the `couldNotBuildFactory` action case should be removed anyway so this might be right, but just want to double check)